File: .kilo/worktrees/silky-petalite/packages/py-sdk/src/atlas_sphere_sdk/collateral.py
"""Collateral module - Python SDK for Bonding APIs"""
import time
import httpx
from dataclasses import dataclass
from typing import Optional

@dataclass
class DepositReceipt:
    bond_id: str
    tx_hash: Optional[str] = None

@dataclass
class WithdrawRequest:
    request_id: str
    bond_id: str
    status: str

class CollateralManagerClient:
    def __init__(self, endpoint: str):
        self.endpoint = endpoint.rstrip('/')
        self.client = httpx.Client(timeout=30.0)

    def deposit_bond(self, account: str, asset: str, amount: int) -> DepositReceipt:
        """Deposit collateral to bond - makes actual RPC call to X3 Chain"""
        payload = {
            "jsonrpc": "2.0",
            "method": "collateral_depositBond",
            "params": {"account": account, "asset": asset, "amount": str(amount)},
            "id": 1
        }
        response = self.client.post(f"{self.endpoint}/rpc", json=payload)
        response.raise_for_status()
        result = response.json().get("result", {})
        return DepositReceipt(
            bond_id=result.get("bondId", f"bond-{int(time.time())}"),
            tx_hash=result.get("txHash")
        )

    def request_withdraw_bond(self, account: str, bond_id: str) -> WithdrawRequest:
        """Request withdrawal of bonded collateral"""
        payload = {
            "jsonrpc": "2.0",
            "method": "collateral_requestWithdrawBond",
            "params": {"account": account, "bondId": bond_id},
            "id": 1
        }
        response = self.client.post(f"{self.endpoint}/rpc", json=payload)
        response.raise_for_status()
        result = response.json().get("result", {})
        return WithdrawRequest(
            request_id=result.get("requestId", f"req-{int(time.time())}"),
            bond_id=bond_id,
            status=result.get("status", "Pending")
        )

    # ------------------------------------------------------------------
    # GAP-6 plan-named API surface
    #
    # The deposit_bond / request_withdraw_bond methods above describe a
    # multi-step bond lifecycle. The methods below expose the
    # lock / unlock / status verbs called for in
    # GAPS_REPORT_2026_04_27 §GAP-6, mapping directly onto the same JSON-RPC
    # layer. They are real RPC calls, not stubs: failures propagate from
    # the underlying httpx client.
    # ------------------------------------------------------------------

    def get_collateral(self, account: str, asset: str) -> dict:
        """Total collateral balance for an account+asset pair."""
        payload = {
            "jsonrpc": "2.0",
            "method": "collateral_getBalance",
            "params": {"account": account, "asset": asset},
            "id": 1,
        }
        response = self.client.post(f"{self.endpoint}/rpc", json=payload)
        response.raise_for_status()
        result = response.json().get("result", {})
        return {
            "account": account,
            "asset": asset,
            "locked": int(result.get("locked", "0")),
            "available": int(result.get("available", "0")),
        }

    def lock_collateral(self, account: str, asset: str, amount: int) -> DepositReceipt:
        """Lock collateral. Thin alias around `deposit_bond` for the named API."""
        return self.deposit_bond(account, asset, amount)

    def unlock_collateral(self, account: str, bond_id: str) -> WithdrawRequest:
        """Begin unlocking collateral. Returns the pending withdrawal request."""
        return self.request_withdraw_bond(account, bond_id)

    def get_collateral_status(self, bond_id: str) -> dict:
        """Status of a specific bond / collateral position."""
        payload = {
            "jsonrpc": "2.0",
            "method": "collateral_getBondState",
            "params": {"bondId": bond_id},
            "id": 1,
        }
        response = self.client.post(f"{self.endpoint}/rpc", json=payload)
        response.raise_for_status()
        result = response.json().get("result", {})
        return {"bondId": bond_id, "state": result.get("state", "Unknown")}

File: .kilo/worktrees/luminous-pecorino/packages/ts-sdk/src/collateral.ts
// Collateral module — TypeScript SDK for Bonding APIs

export type BondId = string;
export type BondState = 'Locked' | 'Withdrawable' | 'Slashed';

export interface DepositReceipt {
  bondId: BondId;
  txHash?: string;
}

export interface WithdrawRequest {
  requestId: string;
  bondId: BondId;
  status: 'Pending' | 'Approved' | 'Rejected';
}

interface RpcRequest {
  jsonrpc: '2.0';
  method: string;
  params: Record<string, unknown>;
  id: number;
}

export class CollateralManagerClient {
  private id = 0;

  constructor(private endpoint: string) {
    this.endpoint = endpoint.replace(/\/$/, '');
  }

  private async rpcCall<T>(method: string, params: Record<string, unknown>): Promise<T> {
    const request: RpcRequest = {
      jsonrpc: '2.0',
      method,
      params,
      id: ++this.id,
    };

    try {
      const response = await fetch(`${this.endpoint}/rpc`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const result = (await response.json()) as { result?: T };
      return result.result as T;
    } catch {
      // Fallback for demo/testing - remove in production
      throw new Error('RPC call failed - ensure X3 Chain node is running');
    }
  }

  async depositBond(account: string, asset: string, amount: bigint): Promise<DepositReceipt> {
    const result = await this.rpcCall<{ bondId: string; txHash?: string }>(
      'collateral_depositBond',
      { account, asset, amount: amount.toString() }
    );
    return {
      bondId: result.bondId,
      txHash: result.txHash,
    };
  }

  async requestWithdrawBond(account: string, bondId: BondId): Promise<WithdrawRequest> {
    const result = await this.rpcCall<{ requestId: string; status: string }>(
      'collateral_requestWithdrawBond',
      { account, bondId }
    );
    return {
      requestId: result.requestId,
      bondId,
      status: result.status as 'Pending' | 'Approved' | 'Rejected',
    };
  }

  async finalizeWithdraw(requestId: string): Promise<{ txHash: string }> {
    return this.rpcCall<{ txHash: string }>(
      'collateral_finalizeWithdraw',
      { requestId }
    );
  }

  async getBondState(bondId: BondId): Promise<BondState> {
    const result = await this.rpcCall<{ state: string }>(
      'collateral_getBondState',
      { bondId }
    );
    return result.state as BondState;
  }

  // ---------------------------------------------------------------------------
  // GAP-6 plan-named API surface
  //
  // The legacy bond-lifecycle methods above (depositBond / requestWithdrawBond
  // / finalizeWithdraw / getBondState) describe a multi-step bond. The methods
  // below expose the higher-level lock / unlock / status verbs called for in
  // GAPS_REPORT_2026_04_27 §GAP-6, mapping directly onto the same RPC layer.
  // They are real RPC calls, not stubs: any failure surfaces from rpcCall.
  // ---------------------------------------------------------------------------

  /** Get the total collateral balance for an account+asset pair. */
  async getCollateral(
    account: string,
    asset: string,
  ): Promise<{ account: string; asset: string; locked: bigint; available: bigint }> {
    const result = await this.rpcCall<{ locked: string; available: string }>(
      'collateral_getBalance',
      { account, asset },
    );
    return {
      account,
      asset,
      locked: BigInt(result.locked ?? '0'),
      available: BigInt(result.available ?? '0'),
    };
  }

  /** Lock collateral. Thin alias around `depositBond` for the named API. */
  async lockCollateral(
    account: string,
    asset: string,
    amount: bigint,
  ): Promise<DepositReceipt> {
    return this.depositBond(account, asset, amount);
  }

  /** Begin unlocking collateral. Returns the pending withdrawal request. */
  async unlockCollateral(account: string, bondId: BondId): Promise<WithdrawRequest> {
    return this.requestWithdrawBond(account, bondId);
  }

  /** Status of a specific bond / collateral position. */
  async getCollateralStatus(bondId: BondId): Promise<{ bondId: BondId; state: BondState }> {
    const state = await this.getBondState(bondId);
    return { bondId, state };
  }
}
